pydantic_ai_slim/pydantic_ai/agent.py

```diff
@@ -1058,3 +1111,181 @@ def _prepare_result_schema(
+
+
+@dataclasses.dataclass
+class AgentRun(Generic[AgentDepsT, ResultDataT]):
+    """A stateful, iterable run of an agent."""
+
+    _graph_run: GraphRun[
+        _agent_graph.GraphAgentState, _agent_graph.GraphAgentDeps[AgentDepsT, Any], MarkFinalResult[ResultDataT]
+    ]
+
+    @property
+    def final_result(self) -> AgentRunResult[AgentDepsT, ResultDataT]:
+        graph_run_result = self._graph_run.final_result
+        return AgentRunResult(graph_run_result)
+
+    def __aiter__(
+        self,
+    ) -> AsyncIterator[
+        BaseNode[
+            _agent_graph.GraphAgentState,
+            _agent_graph.GraphAgentDeps[AgentDepsT, Any],
+            MarkFinalResult[ResultDataT],
+        ]
+        | End[MarkFinalResult[ResultDataT]]
+    ]:
+        return self
+
+    async def __anext__(
+        self,
+    ) -> (
+        BaseNode[
+            _agent_graph.GraphAgentState,
+            _agent_graph.GraphAgentDeps[AgentDepsT, Any],
+            MarkFinalResult[ResultDataT],
+        ]
+        | End[MarkFinalResult[ResultDataT]]
+    ):
+        """Use the last returned node as the input to `Graph.next`."""
+        return await self._graph_run.__anext__()
+
+    async def next(
+        self,
+        node: BaseNode[
+            _agent_graph.GraphAgentState,
+            _agent_graph.GraphAgentDeps[AgentDepsT, Any],
+            MarkFinalResult[ResultDataT],
+        ],
+    ) -> (
+        BaseNode[
+            _agent_graph.GraphAgentState,
+            _agent_graph.GraphAgentDeps[AgentDepsT, Any],
+            MarkFinalResult[ResultDataT],
+        ]
+        | End[MarkFinalResult[ResultDataT]]
+    ):
+        # TODO: It would be nice to expose a synchronous interface for this, to be able to
+        #   synchronously iterate over the agent graph. I don't think this would be hard to do,
+        #   but I'm having a hard time coming up with an API that fits nicely along side the current `run_sync`.
+        #   The use of `await` provides an easy way to signal that you just want the result, but it's less
+        #   clear to me what the analogous thing should be for synchronous code.
+        return await self._graph_run.next(node)
+
+    def usage(self) -> _usage.Usage:
+        return self._graph_run.state.usage
+
+    def get_graph_ctx(
+        self,
+    ) -> GraphRunContext[_agent_graph.GraphAgentState, _agent_graph.GraphAgentDeps[AgentDepsT, Any]]:
+        return GraphRunContext[_agent_graph.GraphAgentState, _agent_graph.GraphAgentDeps[AgentDepsT, Any]](
+            self._graph_run.state, self._graph_run.deps
+        )
+
+
+@dataclasses.dataclass
+class AgentRunResult(Generic[AgentDepsT, ResultDataT]):
+    """The final result of an agent run."""
+
+    graph_run_result: GraphRunResult[
+        _agent_graph.GraphAgentState, _agent_graph.GraphAgentDeps[AgentDepsT, Any], MarkFinalResult[ResultDataT]
+    ]
+
+    @property
+    def result(self) -> MarkFinalResult[ResultDataT]:
```

Comment:
I think we should rename `MarkFinalResult` to `FinalResul` now it's more public, also presumably `.result` here could be private as it didn't exist before.